consensus/core/src/mass/mod.rs

```diff
@@ -66,3 +67,271 @@ pub fn transaction_output_estimated_serialized_size(output: &TransactionOutput)
+    }
+
+    pub fn new_with_consensus_params(consensus_params: &Params) -> Self {
+        Self {
+            mass_per_tx_byte: consensus_params.mass_per_tx_byte,
+            mass_per_script_pub_key_byte: consensus_params.mass_per_script_pub_key_byte,
+            mass_per_sig_op: consensus_params.mass_per_sig_op,
+            storage_mass_parameter: consensus_params.storage_mass_parameter,
+        }
+    }
+
+    /// Calculates the compute mass of this transaction. This does not include the storage mass calculation below which
+    /// requires full UTXO context
+    pub fn calc_tx_compute_mass(&self, tx: &Transaction) -> u64 {
+        if tx.is_coinbase() {
+            return 0;
+        }
+
+        let size = transaction_estimated_serialized_size(tx);
+        let mass_for_size = size * self.mass_per_tx_byte;
+        let total_script_public_key_size: u64 = tx
+            .outputs
+            .iter()
+            .map(|output| 2 /* script public key version (u16) */ + output.script_public_key.script().len() as u64)
+            .sum();
+        let total_script_public_key_mass = total_script_public_key_size * self.mass_per_script_pub_key_byte;
+
+        let total_sigops: u64 = tx.inputs.iter().map(|input| input.sig_op_count as u64).sum();
+        let total_sigops_mass = total_sigops * self.mass_per_sig_op;
+
+        mass_for_size + total_script_public_key_mass + total_sigops_mass
+    }
+
+    /// Calculates the storage mass for this populated transaction.
+    /// Assumptions which must be verified before this call:
+    ///     1. All output values are non-zero
+    ///     2. At least one input (unless coinbase)
+    ///
+    /// Otherwise this function should never fail.
+    pub fn calc_tx_storage_mass(&self, tx: &impl VerifiableTransaction, version: Kip9Version) -> Option<u64> {
+        calc_storage_mass(
+            tx.is_coinbase(),
+            tx.populated_inputs().map(|(_, entry)| entry.amount),
+            tx.outputs().iter().map(|out| out.value),
+            version,
+            self.storage_mass_parameter,
+        )
+    }
+
+    /// Calculates the overall mass of this transaction, combining both compute and storage masses.
+    /// The combination strategy depends on the version passed.
+    pub fn calc_tx_overall_mass(
+        &self,
+        tx: &impl VerifiableTransaction,
+        cached_compute_mass: Option<u64>,
+        version: Kip9Version,
+    ) -> Option<u64> {
+        match version {
+            Kip9Version::Alpha => self
+                .calc_tx_storage_mass(tx, version)
+                .and_then(|mass| mass.checked_add(cached_compute_mass.unwrap_or_else(|| self.calc_tx_compute_mass(tx.tx())))),
+            Kip9Version::Beta => self
+                .calc_tx_storage_mass(tx, version)
+                .map(|mass| mass.max(cached_compute_mass.unwrap_or_else(|| self.calc_tx_compute_mass(tx.tx())))),
+        }
```

Comment:
Consider:

```
        let storage_mass_result = self.calc_tx_storage_mass(tx, version);
        let tx_mass = cached_compute_mass.unwrap_or_else(|| self.calc_tx_compute_mass(tx.tx()));
        match version {
            Kip9Version::Alpha => storage_mass_result.and_then(|mass| mass.checked_add(tx_mass)),
            Kip9Version::Beta => storage_mass_result.map(|mass| mass.max(tx_mass)),
        }
```

Also, why is one `.and_then` and the other `.map`?

Reply:
`map` transforms the value by applying a transform function on the underlying value if it is `Some`, retaining the transformed value afterwards.  whereas `and_then` pretty much does the same thing, but expects the transform output to be an option as well, allowing the transform fn to also transform the output to `None`.

since using `checked_add` here, which returns `Option` `and_then` is more fitting.  if a `saturating_add` was used, then a `map` would be used.  In this context both are sort of the same. They are just more "fitted" to the data pipeline.